aiorwlock/__init__.py

```diff
@@ -157,22 +157,18 @@ def _wake_up(self) -> None:
         # waiters.
         if self._r_state == 0 and self._w_state == 0:
             if self._write_waiters:
-                self._wake_up_first(self._write_waiters)
-            elif self._read_waiters:
-                self._wake_up_all(self._read_waiters)
-
-    def _wake_up_first(self, waiters: 'Deque[Future[None]]') -> None:
-        # Wake up the first waiter who isn't cancelled.
-        for fut in waiters:
-            if not fut.done():
-                fut.set_result(None)
```

Comment:
Problem here is that we set the future to wake up (thus telling it that it has the lock), but it is not run immediately. Therefore, it is possible for a new task to acquire the lock, which will see `_w_state == 0` and then proceed with the lock. Then the first task will resume also thinking it has the lock.

If we increment `_w_state` immediately at the same time as setting the future result, then the new task will see `_w_state !=0` and create a new waiter.